feat: Fix combineSkeletons

0b5vr opened the pull request:
This PR fixes `combineSkeletons`.

There is a model that uses the same skinIndex accessor but not the same skinWeight accessors between meshes.
This PR updates the implementation of attributeUsedIndexSetMap to use both skinIndex and skinWeight accesstors.

The same skinIndex accessor could also be used between different bone sets.
Clone the bone set if the skinIndex accessor is the same but the bone set is different.

0b5vr committed:
feat: Fix combineSkeletons

There is a model that uses the same skinIndex accessor but not the same skinWeight accessors between meshes
This commit updates the implementation of attributeUsedIndexSetMap to use both skinIndex and skinWeight accesstors

The same skinIndex accessor could also be used between different bone sets
Clone the bone set if the skinIndex accessor is the same but the bone set is different

0b5vr commented on packages/three-vrm/src/VRMUtils/combineSkeletons.ts:
```diff
@@ -222,3 +235,62 @@ function matrixEquals(a: THREE.Matrix4, b: THREE.Matrix4, tolerance?: number) {
+  // Collect skin index attributes
+  // skinIndex attribute might be shared among different bone sets
+  // If there are multiple bone sets that share the same skinIndex attribute, clone the attribute
+  const skinIndexNewSkinIndexBonesMapMap = new Map<
```
If there is a skin index accessor that is used with two or more different bone sets, you can check by seeing inside the `skinIndexNewSkinIndexBonesMapMap`

ke456-png commented on packages/three-vrm/src/VRMUtils/combineSkeletons.ts:
```diff
@@ -222,3 +235,62 @@ function matrixEquals(a: THREE.Matrix4, b: THREE.Matrix4, tolerance?: number) {
+      continue;
+    }
+
+    // Check if the bone set is already registered
+    for (const bones of newSkinIndexBonesMap.values()) {
+      if (arrayEquals(bones, mesh.skeleton.bones)) {
+        continue;
+      }
+    }
```
There is a problem with the scope of the `continue`.

0b5vr replied:
oh no

0b5vr replied:
fixed @ https://github.com/pixiv/three-vrm/pull/1569/commits/902e6d83a4048a1e7a6bf4dfdb51252d7f8eb456

I also forgot to set the new attribute when we found an already registered skin index attribute 😅

0b5vr committed:
fix: Fix combineSkeletons, collectSkinIndexAttrs

The continue statement was not working as intended

I also forgot to set the new attribute when we found an already registered skin index attribute

I had to change eslint `@typescript-eslint/naming-convention` rule in order to use `_` for parameter name
See: https://typescript-eslint.io/rules/naming-convention/

ke456-png reviewed (approved):
Thanks!
